optuna/samplers/_gp/_sampler.py

```diff
@@ -0,0 +1,169 @@
+from optuna.study import Study
+from optuna.study import StudyDirection
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+if TYPE_CHECKING:
+    import torch
+
+    import optuna._gp._acqf as _acqf
+    import optuna._gp._gp as _gp
+    import optuna._gp._optim as _optim
+    import optuna._gp._search_space as _search_space
+else:
+    torch = _LazyImport("torch")
+    _search_space = _LazyImport("optuna._gp._search_space")
+    _gp = _LazyImport("optuna._gp._gp")
+    _optim = _LazyImport("optuna._gp._optim")
+    _acqf = _LazyImport("optuna._gp._acqf")
+
+
+def default_log_prior(kernel_params: "_gp.KernelParams") -> "torch.Tensor":
+    def gamma_log_prior(x: "torch.Tensor", concentration: float, rate: float) -> "torch.Tensor":
+        return (concentration - 1) * torch.log(x) - rate * x
+
+    return (
+        gamma_log_prior(kernel_params.inv_sq_lengthscales, 2, 0.5).sum()
+        + gamma_log_prior(kernel_params.kernel_scale, 2, 1)
+        + gamma_log_prior(kernel_params.noise, 1.1, 20)
+    )
+
+
+@experimental_class("3.6.0")
+class GPSampler(BaseSampler):
+    """Sampler using Gaussian process-based Bayesian optimization.
+
+    This sampler fits a Gaussian process (GP) to the objective function and optimizes
+    the acquisition function to suggest the next parameters.
+
+    The current implementation uses:
+    - Matern kernel with nu=2.5 (differentiable twice)
+    - Automatic relevance determination (ARD) for a lengthscale of each parameter
+    - Gamma prior for lengthscales, kernel scale, and noise scale
+    - Log Expected Improvement (logEI) as the acquisition function
+    - Quasi-Monte Carlo (QMC) sampling to optimize the acquisition function
+
+    .. note::
+        This sampler requires `scipy` and `pytorch` to be installed.
+        You can install these dependencies with `pip install scipy pytorch`.
+
+    Args:
+        seed:
+            Random seed passed to `independent_sampler`.
+            Defaults to None (random seed is used).
+
+        independent_sampler:
+            Sampler used for initial sampling (for the first `n_startup_trials` trials)
+            and for conditional parameters. Defaults to `None` (a random sampler is used).
+
+        n_startup_trials:
+            Number of initial trials. Defaults to 10.
+    """
+
+    def __init__(
+        self,
+        *,
+        seed: int | None = None,
+        independent_sampler: BaseSampler | None = None,
+        n_startup_trials: int = 10,
+    ) -> None:
+        self._independent_sampler = independent_sampler or optuna.samplers.RandomSampler(seed=seed)
+        self._intersection_search_space = optuna.search_space.IntersectionSearchSpace()
+        self._n_startup_trials = n_startup_trials
+        self._log_prior: "Callable[[_gp.KernelParams], torch.Tensor]" = default_log_prior
+        self._minimum_noise: float = 1e-6
+        self._last_kernel_params: "_gp.KernelParams | None" = None
```

Comment:
[nit] Do you think `cache` is better? (At least, I needed to read the corresponding lines to understand what `last` means)

```suggestion
        self._kernel_params_cache: "_gp.KernelParams | None" = None
```